Move judgement result revert logic to Playfield

ekrctb opened the pull request:
Previously, some judgement results were not reverted when the source DHO is not alive (e.g. frames skipped in editor).
Now, all results are reverted in the exact reverse order.
`HitObjectContainer.RevertResult` event is removed and `Playfield.RevertResult` event is changed to remove `DrawableHitObject` argument.

- [x] Depends on #22292
- Fix #21090

ekrctb committed:
Move judgement result revert logic to Playfield
Previously, some judgement results were not reverted
when the source DHO is not alive (e.g. frames skipped in editor).
Now, all results are reverted in the exact reverse order.

ekrctb committed:
Remove now-redundant code

peppy commented:
I think it would be good to add a test here which fails on `master`, so we can better understand the specific fail case (and guard against it potentially regressing in the future).

ekrctb committed:
Add test for RevertResult

bdach commented on osu.Game/Rulesets/UI/Playfield.cs:
```diff
@@ -443,6 +452,25 @@ public DrawableSamplePool(ISampleInfo sampleInfo, int initialSize, int? maximumS
 
+        private void onNewResult(DrawableHitObject drawable, JudgementResult result)
+        {
+            // Not using result.TimeAbsolute because that might change and also there is a potential precision issue.
+            judgementResults.Push(new JudgementResultEntry(Time.Current, drawable.Entry.AsNonNull(), result));
```
Do you mind elaborating on this? What might change? What precision issue?

Sounds to me like you may be referring to what I already found as https://github.com/ppy/osu/issues/17482#issuecomment-1079986449? If so, then using `Time.Current` as a local workaround and leaving `TimeAbsolute` as a footgun for the next person to blow their foot off with seems pretty wrong to me.

ekrctb replied:
Change of `HitObject.StartTime`.
Precision issue is the catastrophic cancellation in `TimeAbsolute = StartTime + (hit time - StartTime)` so it might be slightly different from the original `hit time`.
On a second thought it might be better to fix `TimeAbsolute` issue by storing the original hit time in it rather than duplicating here. Though that is a potential behavior change so I'll try to see if it breaks something.

bdach commented on osu.Game/Rulesets/UI/Playfield.cs:
```diff
@@ -443,6 +452,25 @@ public DrawableSamplePool(ISampleInfo sampleInfo, int initialSize, int? maximumS
+        private void revertResult(JudgementResultEntry entry)
+        {
+            var result = entry.Result;
+            RevertResult?.Invoke(result);
+
+            result.TimeOffset = 0;
+            result.Type = HitResult.None;
+
+            entry.HitObjectEntry.OnRevertResult();
```
The fact that `Playfield.RevertResult` event subscribers see `result` with everything populated, but `HitObjectLifetimeEntry.RevertResult` subscirber (and therefore, consequently, `DrawableHitObject.RevertResult` subscribers) are seeing the result with time offset and type cleared scares me a lot. My confidence that this is correct is basically zero unless I go and chase down all event pathways. Is this preserving existing behaviour?

ekrctb replied:
Good catch. This was not intentional, no.

bdach commented on osu.Game/Rulesets/Judgements/JudgementResultEntry.cs:
```diff
@@ -0,0 +1,26 @@
+            Result = result;
+        }
+
+        public int CompareTo(JudgementResultEntry? other) => Time.CompareTo(other?.Time);
```
This has boxing allocation overhead (overload used is `CompareTo(object?)`). With something as important as this maybe important to hand-roll the null check to avoid it.

ekrctb replied:
Noticed that this comparator is not actually used.
(stray code from when I was trying to use an ordered set instead of the current stack)

ekrctb committed:
Remove unused code

ekrctb committed:
Fix OnRevertResult timing

ekrctb committed:
Store Result.TimeAbsolute separately from offset
Calculating from TimeOffset is bad because it loses precision.
The result time won't change anymore
even If `HitObject.GetEndTime()` changes later.

ekrctb committed:
Fix inspection issue

bdach commented on osu.Game/Rulesets/Judgements/JudgementResult.cs:
```diff
@@ -33,16 +33,19 @@ public class JudgementResult
+        /// <remarks>
+        /// This is initially set to the end time of <see cref="HitObject"/>.
+        /// </remarks>
+        public double TimeAbsolute { get; internal set; }
```
Am I the only one that's completely spooked that `TimeOffset` and `TimeAbsolute` are both independently settable with this, and therefore can completely desync from one another?

I've written [a branch](https://github.com/ppy/osu/compare/master...bdach:dho-store-absolute-time-2?expand=1) that would have both settable before, but with the caveat that setting one would also propagate to the other. Inverting the relationship and having `TimeOffset` be derived from `TimeAbsolute` [was also floated at one point](https://github.com/ppy/osu/issues/17482#issuecomment-1084351300), but that [is more work](https://github.com/ppy/osu/issues/17482#issuecomment-1084365647).

ekrctb replied:
I also concerned about that, but concluded this is okay for now because it is `internal set`.
Ideally we don't want to use such `internal` properties for such a distant coupling IMO but it seems difficult to change it without much refactoring.

bdach replied:
@peppy just to confirm, did you see this when reviewing prior to approving? Like to me this is very scary, even if it's internal. We can also break this too.

Would like @smoogipoo's opinion on this too while we're at it.

peppy replied:
Yep. I'm quite fine with this. I don't mind cases like this because it's hard coupled at one or two points of interaction. I have faith that if that ever changed one of us would raise alarm bells and we'd reconsider at that point.

bdach replied:
Okay then. In that case, I self requested re-review as I will want to look through all cases where either property is set to make sure that the other doesn't contain a value that's invalid for reading.

peppy commented:
@smoogipoo interested in your high level thoughts on this change as well.

peppy committed:
Merge branch 'master' into revert-result-in-playfield

peppy committed:
Update test to use newer assetion logic

peppy commented on osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs:
```diff
@@ -578,26 +587,6 @@ public virtual void StopAllSamples()
-        {
-            base.Update();
-
-            if (Result != null && Result.HasResult)
```
This is a huge win.

peppy commented on osu.Game/Rulesets/Objects/HitObjectLifetimeEntry.cs:
```diff
@@ -26,6 +27,8 @@ public class HitObjectLifetimeEntry : LifetimeEntry
 
         private readonly IBindable<double> startTimeBindable = new BindableDouble();
 
+        internal event Action? RevertResult;
```
This is a touch awkward in terms of event flow, but I can't immediately see a way out of it.

peppy reviewed (dismissed):
I *think* I am on board with this change.

bdach commented on osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs:
```diff
@@ -684,7 +673,8 @@ protected void ApplyResult(Action<JudgementResult> application)
             }
 
-            Result.TimeOffset = Math.Min(MaximumJudgementOffset, Time.Current - HitObject.GetEndTime());
+            Result.TimeAbsolute = Time.Current;
+            Result.TimeOffset = Math.Min(MaximumJudgementOffset, Result.TimeAbsolute - HitObject.GetEndTime());
```
See this sort of thing is what screws me up about having both times settable independently. `TimeOffset` is clamped to `MaximumJudgementOffset`, but `TimeAbsolute` is stored raw without so much as looking at that property?

So if `ApplyResult()` is called after the maximum judgement offset is up, you can potentially have a situation wherein `HitObject.GetEndTime() + TimeOffset != TimeAbsolute` (fundamentally and mathematically, putting FP shenanigans aside), because only one of the two gets the clamp and `TimeAbsolute` is allowed to be *any* value, even potentially in excess of `HitObject.GetEndTime() + MaximumJudgementOffset`?

I really *really* do not like this. I won't block the PR over it if it's just me, but really this to me is the stuff of nightmares.

smoogipoo replied:
What is the unclamped `TimeAbsolute` useful for? In other words, the purpose of the change was to decouple the property from the hitobject's start time, but this would still be true if the clamping where added back.

ekrctb replied:
This is bad and probably regress some cases described in the original `MaximumJudgementOffset` introduction PR #10962.
With #22292, it has access to `MaximumJudgementOffset` here so:

```csharp
internal double TrueTime;

double TimeOffset => Min(TrueTime - HitObject.GetEndTime(), HitObject.MaximumJudgementOffset);
double TimeAbsolute => Min(TrueTime, HitObject.GetEndTime() + HitObject.MaximumJudgementOffset);
```

Looks okay? (not sure about naming of `TrueTime`)

bdach replied:
The proposal above with one source-of-truth internal property is sitting much better with me, but it will not be as simple as that - there are test usages that will need updating.

But, as @smoogipoo already asked - do we *really* need `TrueTime` for anything? Maybe we can just set the proper value for now, and apply the proposal above separately in another PR?

ekrctb replied:
Here, `TrueTime` is needed to handle revert logic by a stack (last-in-first-out).
Alternatively, the set of judged results can be maintained by a sorted set and that will be unintuitive that: A judged then B judged, but A-result is reverted first then B-result reverted.

bdach replied:
I see.

I'm ok with `TrueTime` (maybe `RawTime` is better? likely needs xmldoc either way), but applying that won't be straightforward. There are test usages that set `TimeOffset` directly which will need to be migrated over.

peppy replied:
What's the go here? Are we going to rename the parameter still? I'd agree with `RawTime` more than `TrueTime`.

bdach replied:
Uh, I'm okay with adding `TrueTime` or `RawTime` as proposed, it just needs to be actually _applied_ here. And I don't see any changes for it in this PR, and as I said, it is not going to be as simple as the three lines above.

peppy replied:
Safe to say that that means things here are probably a tad too complex and I'm missing what is actually being proposed.

Feels like we need to take a step back and reconsider.

smoogipoo commented:
At a high level, this looks fine to me. The concerns above are quite valid though.

ekrctb commented:
Separate question, should `MaximumJudgementOffset` be applied for the negative direction also? It makes more sense if this property is used for lifetime purpose #22292.

ekrctb committed:
Remove JudgementResultEntry
It is not needed anymore as TimeAbsolute is stored raw.

bdach commented:
> Separate question, should `MaximumJudgementOffset` be applied for the negative direction also?

I had that thought too when reading the code, but refrained from pointing it out because I didn't want to scope creep this PR.

ekrctb committed:
Merge branch 'maximum-judgement-offset-in-hit-object' into revert-result-in-playfield

ekrctb committed:
Store RawTime in JudgementResult

ekrctb committed:
Fix test setting invalid TimeOffset

peppy commented on osu.Game/Rulesets/Judgements/JudgementResult.cs:
```diff
@@ -33,16 +34,30 @@ public class JudgementResult
+        /// <remarks>
+        /// This is used instead of <see cref="TimeAbsolute"/> to check whether this <see cref="JudgementResult"/> should be reverted.
+        /// </remarks>
+        internal double? RawTime { get; set; }
```
In terms of naming, my only alternative proposal here would be

`RawTime` -> `Time`
`TimeAbsolute` -> `ClampedTime`

Both have roughly the same number of usages, so I'm not sure about which should get a non-suffixed/prefixed name. But in general I dislike naming with words that have unclear meanings (like `Raw` or `Base` or `Real`).

bdach replied:
`ClampedTime` is pretty ugly name for a public property to have. I'd argue the reverse and keep `public TimeAbsolute` and `internal UnclampedTime`. Rationale is that the ugly name is contained to internal members I suppose?

ekrctb replied:
It is not clear having clamped time in `JugementResult` should be the way forward (why do we have judgements with out-of-bound time offsets in the first place?) so let's not name it `ClampedTime` for now.

peppy replied:
Sure, i'm okay with no change if that's deemed better. We can see how it plays out.

peppy reviewed (commented):
Apart from naming, I'm pretty happy with this direction now.

bdach committed:
Merge branch 'master' into revert-result-in-playfield

bdach commented:
I can agree with what this pull is doing, I don't see much wrong in what it's doing, and it will also fix other issues in the process (like https://github.com/ppy/osu/issues/17482). Doesn't make the change any less scary, mind, but yeah. Way I see it it'll all either go perfectly or we'll have to revert this in a week. But probably worth a go regardless?

bdach committed:
Merge branch 'master' into revert-result-in-playfield
